Tooltip redux

derekeder opened the pull request:
## Overview

Updates the map interface to support click interactions and now saves state in the URL for easier sharing. The popups allow text to be easier to read and copy/paste over the hover, that updates whenever the mouse moves.

- clicking a feature shows a popup tooltip
- hovering over a feature updates a simple info box in the lower left corner
- clicking a feature saves the ID to the URL. loading from that URL clicks the feature on load

Related to #78. This likely eliminates/reduces the need for creating detail pages.

Examples:

* https://deploy-preview-82--il-solar-map.netlify.app/#/?id=17045070200
* https://deploy-preview-82--il-solar-map.netlify.app/#/?category=utility_kw&status=planned&id=17019010800
* https://deploy-preview-82--il-solar-map.netlify.app/#/?category=total_kw&status=energized&geography=places&id=53886039

### Demo

![Screenshot 2025-05-01 at 4 59 27 PM](https://github.com/user-attachments/assets/7b789537-50d3-4016-a25c-1a9421010c9d)

## Testing Instructions

* toggle hover and click on the map and confirm features work as expected
* reload the page and confirm the selected feature is still shown
* save a few URLs and confirm it loads the selected feature and layer

derekeder commented:
update: now showing the hover outlines more clearly. this required adding outline-specific layers for each geography and styling accordingly:
![Screenshot 2025-05-04 at 5 18 05 PM](https://github.com/user-attachments/assets/79fca301-1d8b-4b54-9094-cbb62b2b4f66)

vkoves commented on css/custom.css:
```diff
@@ -34,6 +34,24 @@ body, p {
+    margin: 5px;
+    font: 14px/16px Arial, Helvetica, sans-serif;
+    background: white;
+    /*background: rgba(255,255,255,0.8);*/
```
This should be deleted

vkoves commented on js/map.js:
```diff
@@ -322,4 +432,29 @@ map.on('load', () => {
   })
+
+  // if an ID is in the URL, load it
+  map.on('sourcedata', function (e) {
```
Might be worth noting what this `e` event type is, for future explorers

vkoves commented on js/map.js:
```diff
@@ -49,28 +49,59 @@ const geoButtonsSelector = '#geography-select button'
+    this.map = undefined
+  }
+
+  updateInfo(content) {
```
I wish this was Typescript, but since it's not, typed doc blocks might be helpful (e.g. `/** @param content {string} The CSS display style to use */`)

vkoves reviewed (changes requested):
Have a few minor code comments, but the outline seems a little finicky.

1. It doesn't show on _just clicking_ (you have to reload)

![image](https://github.com/user-attachments/assets/fc11a323-94a3-4962-a0d6-6c1d3f44cdcc)

2. It goes away when you hover over something else, even though the tooltips stays.

I'd consider making a distinct active and hover state to make this a little cleaner - this could be adding a background effect to the active district as well, making the outline persist and thicker, etc.

But also it's your baby, so feel free to ignore all my thoughts!

derekeder commented on js/map.js:
```diff
@@ -229,14 +310,30 @@ function addLayer(map, layerSource, visible = 'none'){
+      { source: layerSource, id: hoveredPolygonId },
+      { hover: false }
+    )
+    infoControl.updateInfo('Hover over a feature to see details')
```
good catch!

derekeder commented on js/map.js:
```diff
@@ -294,6 +398,12 @@ map.on('load', () => {
     map.setLayoutProperty('il-senate-fills', 'visibility', 'none')
     map.setLayoutProperty('il-house-fills', 'visibility', 'none')
+
+    map.setLayoutProperty('tracts-outline', 'visibility', 'none')
```
yup we can tighten this up

derekeder commented:
@vkoves thx! I addressed the code comments and updated the hover/clicked logic and styles to be independent.

On hover, the opacity is slightly increased. On click, the black outline is shown. Here's a demo:

https://github.com/user-attachments/assets/931196a6-1954-44c7-bad8-53dfb01b2950

vkoves reviewed (approved):
This looks great now, awesome stuff!

One side thing I wanted to mention is that it'd be nice to have a back button that snaps you back to the state level view - at least for me it's a bit annoying to zoom all the way back out after clicking into a small tract, but 🤷🏻‍♂️

derekeder commented:
@vkoves thx! i had a similar thought. I'll open an issue and add it in another PR
